### PR TITLE
implement basic buy and sell policies

### PR DESCRIPTION
**Pull request by rwcarlsen**
The only thing really missing here is tests.  I'm not sure of a *good* way to effectively test this API other than creating a full fake mock resource exchange that calls these functions and carefully checks through returned portfolios, etc.  I would be happy to have help writing the tests.  I won't have much time soon ;-)

### CODE REVIEW

**Comment by gidden** [comment]
Hi @rwcarlsen, I'm likely going to be looking at similar things to this PR, and I think this is a great start. Is the testing issue overcome by your mock sim work?

**Comment by rwcarlsen** [comment]
mocksim should help a lot for testing here.  You are welcome to pick this
up.  I won't have time for it in the near term.

File: src/toolkit/buy_policy.cc [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+BuyPolicy& BuyPolicy::Set(std::string commod, Composition::Ptr c, double pref) {
+  CommodDetail d;
+  d.comp = c;
+  d.pref = pref;
+  commods_[commod] = d;
+  return *this;
+}
+
+void BuyPolicy::Start() { manager()->context()->RegisterTrader(this); }
+
```
Comment (mbmcgarry) [inline src/toolkit/buy_policy.cc]: shouldn't this line be indented?

File: src/toolkit/buy_policy.cc [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+void BuyPolicy::Stop() { manager()->context()->UnregisterTrader(this); }
+
```
Comment (mbmcgarry) [inline src/toolkit/buy_policy.cc]: same as above?

File: src/toolkit/buy_policy.cc [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  bool exclusive = true;
+  if (quantize_ < 0) {
+    exclusive = false;
+    quanta = amt;
+  }
+
+  LGH(INFO3) << "requesting " << amt << " kg";
+
+  RequestPortfolio<Material>::Ptr port(new RequestPortfolio<Material>());
+  std::map<std::string, CommodDetail>::iterator it;
```
Comment (mbmcgarry) [inline src/toolkit/buy_policy.cc]: indent?

File: src/toolkit/buy_policy.cc [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  double quanta = quantize_;
+  bool exclusive = true;
+  if (quantize_ < 0) {
+    exclusive = false;
+    quanta = amt;
+  }
+
+  LGH(INFO3) << "requesting " << amt << " kg";
+
+  RequestPortfolio<Material>::Ptr port(new RequestPortfolio<Material>());
```
Comment (mbmcgarry) [inline src/toolkit/buy_policy.cc]: Indent?

File: src/toolkit/sell_policy.cc [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      if (quantize_ < 0) {
+        Material::Ptr m = buf_->Pop<Material>();
+        buf_->Push(m);
+        Material::Ptr offer = Material::CreateUntracked(qty, m->comp());
+        port->AddBid(req, offer, this, exclusive);
+        LG(INFO4) << "  - bid " << qty << " kg on a request for " << commod;
+      } else {
+        for (int i = 0; i < (int)(qty / quantize_); i++) {
+          Material::Ptr m = buf_->Pop<Material>();
+          buf_->Push(m);
```
Comment (mbmcgarry) [inline src/toolkit/sell_policy.cc]: Can't you put this outside the if/else since it's doing the same thing regardless of the value of quantize_?

**Comment by mbmcgarry** [comment]
This seems like it will be a nice new feature.  I am confused though as to how preferences are adjusted with this change.  Since AdjustMatlPrefs is now being called inside the buy/sell policy internal to cyclus, where/how are they passed in from a cycamore archetype?  I assume the examples in the archetype dev guide ( http://fuelcycle.org/arche/dre.html#adj ) are no longer applicable in conjunction with the buy/sell policy. Can you provide a new example?

Also there are several places (esp. in buy_policy.cc) where it seems like the indentation isn't following the style guide, although I am admittedly unsure about how to indent when the function definition has to be split onto two lines.

**Comment by mbmcgarry** [comment]
Just saw that this was supposed to be closed by #1127 . Sorry for the confusion. I'll look at that one and see if my comments are addressed already.

**Comment by gidden** [comment]
@rwcarlsen, I'm going to close this in favor of #1127

**Pull request closed without merging**